### PR TITLE
Update application design documentation

### LINKED ISSUE #251: [docs] Update Design Patterns Documentation
The [design pattern doc](https://github.com/grafana/grafana-app-sdk/blob/main/docs/design-patterns.md) is somewhat out-of-date and is rather short. It could use expanding and updated explanations/diagrams based on current app platform ideas.

### PR DESCRIPTION
**Pull request by radiohead**
### WHAT

This commit updates the documentation related to the application design. The new version is more up-to-date, with explanations of the underlying concepts and updated application types.

### WHY

To make sure that our documentation is in line with the current vision and development direction of the app platform and the app SDK.

Closes #251

### CODE REVIEW

**Review by IfSentient — APPROVED** [review]
LGTM, just a few small comments.

Could we also include the source for the excalidraw diagrams, to make updating these docs easier in the future?

**Comment by radiohead** [comment]
> Could we also include the source for the excalidraw diagrams, to make updating these docs easier in the future?

@IfSentient the source is actually already included. Excalidraw has this neat trick where it can embed the scene into the PNG metadata, so those `XXX.excalidraw.png` files can be edited in Excalidraw (both web and / or VSCode extension).

Still, if you want me to include the raw metadata files as well let me know and I'll do that!

File: docs/application-design/platform-concepts.md [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+## Document-oriented unified storage
+
+The platform uses so-called "document-oriented" storage model, which in practice means that the data is stored as key-value pairs, with value consisting of metadata (creation timestamps, tenant IDs, labels and so on) and binary payload data. For example, a dashboard might be stored with its UID as key, its folder as metadata and its JSON model as a binary blob.
+
+The storage exposes a relatively simple (compared to e.g. SQL) API:
+
+* `Get` - returns a single object for a given API group, version and Kind, using object's UID.
+* `List` - fetches a list of objects for a given API group, version and Kind, e.g. all Dashboards. It supports pagination and some basic filtering by namespaces or object labels.
+* `Watch` - works similarly to `List` but instead of returning immediately, instead opens a stream, which receives newly created objects, objects which have been updated or deleted. It lets the client "watch" for changes of a specific Kind, hence the name.
```
[the remark this answers is not in the document]
Comment (radiohead) [reply docs/application-design/platform-concepts.md]: Completely agree, fixed!

File: docs/application-design/platform-concepts.md [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+If we follow this diagram, we can make the following requests to the API:
+
+* `GET /core.grafana.app/v1alpha1/panels` lists all `Panel` resources of the `v1alpha1` version in the `core.grafana.app` API group.
+* `GET /core.grafana.app/v1/namespaces/default/queries/foo/querylimit` will retrieve the `QueryLimit` subresource of the `foo` `Query` resource of the `v1` version in the `core.grafana.app` API group, in the `default` namespace.
+
+Standardizing the API makes it easy to extend it, build tools that integrate with it, generate client code and so on. However, it also requires developers to adhere to it and design their applications keeping in mind some of the API limitations.
+
+Most important limitations are:
+
+* All routes have to be resource-oriented. Think about how you structure your application's logic around resources & subresources. For example, if you want to have a resource with some usage stats, you cannot have an RPC-lime API for it, you have to create a dedicated `FooStats` subresource (assuming your resource is called `Foo`), which would contain a stats object.
```
[the remark this answers is not in the document]
Comment (radiohead) [reply docs/application-design/platform-concepts.md]: Good catch, fixed!

**Comment by IfSentient** [comment]
> > Could we also include the source for the excalidraw diagrams, to make updating these docs easier in the future?
> 
> @IfSentient the source is actually already included. Excalidraw has this neat trick where it can embed the scene into the PNG metadata, so those `XXX.excalidraw.png` files can be edited in Excalidraw (both web and / or VSCode extension).
> 
> Still, if you want me to include the raw metadata files as well let me know and I'll do that!
> 
> I've also addressed your comments – feel free to take another look and merge if it looks good to you. Thanks!

Neat, TIL

**Review by IfSentient — APPROVED** [review]

**Pull request merged**